ADBDEV-4969: Fix resource group tests

KnightMurloc opened the pull request:
Fix resource group tests

Patch 52bbf95d91155517a711b193bbffbaec880e7389 increased memory consumption on
the coordinator and caused the resgroup_cpu_rate_limit test to fail.

This patch adds the ability to create a demo cluster on multiple hosts, which
allows us to create an environment close to the upstream.
Also disables the ORCA planner for resource group tests for two reasons. First,
starting with commit 39de39eb6edb0d0f94f533e96e04b66d6c790d14, upstream runs
tests without it. Secondly, in the resgroup_cpu_rate_limit test, there is a
problem that when the CPU is limited, ORCA can plan a query for a long time,
which leads to the test failing.

KnightMurloc committed:
Fix resource group tests

Patch 52bbf95d91155517a711b193bbffbaec880e7389 increased memory consumption on
the coordinator and caused the resgroup_cpu_rate_limit test to fail.

This patch adds the ability to create a demo cluster on multiple hosts, which
allows you to create an environment close to the upstream.
Add a script to run resource group tests.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66056

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1132144

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66059

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1132386

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1132387

RekGRpth commented on arenadata/scripts/run_resgroup_test.bash:
```diff
@@ -0,0 +1,104 @@
+docker network create resgroup_net
+
+#start cdw and sdw1 containers
+docker run -d --rm \
```
why not use docker compose like in behave tests?

KnightMurloc replied:
It is possible, but I thought that since we only need two containers, it would be superfluous.

RekGRpth replied:
If you run tests for resource groups in the same way as behave tests through Docker Compose, then you do not need to modify the demo cluster and this modification can be done in a separate task, because it is still useful and necessary.

KnightMurloc replied:
Technically, it is possible to use containers for behave tests to run tests of resource groups but. We will need to put the initialization of containers in a separate script to avoid duplication. We still need the ability to deploy a demo cluster on multiple hosts.

RekGRpth replied:
Don't behave tests run a cluster in multiple containers without using the new demo cluster feature?

KnightMurloc replied:
Yes, they use it https://github.com/arenadata/gpdb/blob/adb-6.x/gpMgmt/test/behave_utils/cluster_setup.py

RekGRpth replied:
Then, it turns out that with the new feature of the demo cluster it is possible to simplify the behave tests?

KnightMurloc replied:
Maybe. but do we want such big changes?

KnightMurloc replied:
If we want to use docker-compose for behave tests, then it's probably worth waiting for merge https://github.com/arenadata/gpdb/pull/885 to avoid conflicts.

RekGRpth replied:
It was merged

KnightMurloc replied:
So in the end, you think it's worth reusing the containers of the behave tests? Right?

RekGRpth replied:
> the containers of the behave tests

may be not of but like in

KnightMurloc committed:
Put container initialization in a separate script.
So that it can be used in both resource group and behave tests.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66380

RekGRpth commented:
Fix conflict

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66383

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1147595

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1147596

RekGRpth commented on arenadata/scripts/run_resgroup_test.bash:
```diff
@@ -0,0 +1,64 @@
+bash arenadata/scripts/init_containers.sh $project cdw sdw1
+
+#grant access rights to group controllers
+docker-compose -p $project -f arenadata/docker-compose.yaml exec cdw bash -c "
```
Why not use a service loop here?

RekGRpth commented on arenadata/scripts/init_containers.sh:
```diff
@@ -0,0 +1,39 @@
+if [[ $# -eq 0 ]]; then
+  services=$(docker-compose -p $project -f arenadata/docker-compose.yaml config --services | tr '\n' ' ')
+else
+  services="$@"
```
Why can't we just use the entire docker-compose.yaml file with the coordinator and three segments for resource group tests?

KnightMurloc replied:
If we don't care about running extra containers, then we can.

RekGRpth replied:
I think this will be closer to reality.

KnightMurloc committed:
use loop to init cgroups in containers. up all containers.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66507

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1153331

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1153326

RekGRpth commented:
Why did ORCA regression tests fail?

RekGRpth commented:
> Why did ORCA regression tests fail?

it is failed even on local run
```sh
./pg_isolation2_regress --init-file=../../../src/test/regress/init_file --init-file=./init_file_isolation2 --load-extension=gp_inject_fault segwalrep/fts_unblock_primary segwalrep/recoverseg_from_file
```

KnightMurloc commented:
I was unable to reproduce the problem locally.

RekGRpth commented:
> I was unable to reproduce the problem locally.

Surprisingly, now I can’t reproduce either! Maybe these are new flaky tests? Then add them to the wiki and create tasks.

andr-sokolov commented on gpAux/gpdemo/demo_cluster.sh:
```diff
@@ -247,6 +252,30 @@ if [ ! -x $GPPATH/gpinitsystem ]; then
+HOSTS_OPTS=""
+TRUSTED_SHELL="`pwd`/lalshell"
+rm -f hostfile
+if [[ -v HOSTS_LIST  ]]; then
```
Is it ok, when HOSTS_LIST is empty string?

andr-sokolov commented on arenadata/scripts/init_containers.sh:
```diff
@@ -0,0 +1,33 @@
+
+shift
+
+docker-compose -p $project -f arenadata/docker-compose.yaml --env-file arenadata/.env up -d
```
Why we run all containers (4) when we need 2 only?

andr-sokolov commented on arenadata/scripts/run_resgroup_test.bash:
```diff
@@ -0,0 +1,61 @@
+            ${CONFIGURE_FLAGS}
+
+        make -C /home/gpadmin/gpdb_src/src/test/regress
+        ssh sdw1 mkdir -p /home/gpadmin/gpdb_src/src/test/regress </dev/null
```
is it possible to remove "/dev/null"?

KnightMurloc replied:
`</dev/null` is needed to detach `stdin` from the process and pass `EOF` to it when trying to read from it. without this, for some reason, the script does not execute after the ssh command.

KnightMurloc committed:
cosmetic fixes in demo_cluster.sh , removed the creation of ssh keys from the
script to run behave tests as they are created in init_containers.sh The ability
to create not all containers has been returned.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66894

BenderArenadata commented:
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175472

BenderArenadata commented:
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175470

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175479

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175477

RekGRpth commented:
@KnightMurloc, update branch and fix tests, please

andr-sokolov commented on gpAux/gpdemo/demo_cluster.sh:
```diff
@@ -247,8 +250,33 @@ if [ ! -x $GPPATH/gpinitsystem ]; then
+
 if [ -d $DATADIRS ]; then
   rm -rf $DATADIRS
+  gpssh -f hostfile rm -rf $DATADIRS
```
Do we need to remove DATADIRS on servers from hostfile, when $DATADIRS doesn't exist on local host?

andr-sokolov commented on gpAux/gpdemo/demo_cluster.sh:
```diff
@@ -247,8 +250,33 @@ if [ ! -x $GPPATH/gpinitsystem ]; then
+
+LOCALHOST=`hostname`
+
+if [ -n HOSTS_LIST  ]; then
```
redundant space before  ]

BenderArenadata commented:
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175471

BenderArenadata commented:
Failed job Regression tests with ORCA on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175473

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175480

BenderArenadata commented:
Failed job Behave tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1175478

KnightMurloc committed:
Fix demo_cluster.sh

KnightMurloc committed:
always remove data directories on hosts

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/66932

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1177813

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1177814

BenderArenadata commented:
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1177811

RekGRpth commented on arenadata/scripts/run_resgroup_test.bash:
```diff
@@ -0,0 +1,60 @@
+        scp /home/gpadmin/gpdb_src/src/test/regress/regress.so \
+            gpadmin@sdw1:/home/gpadmin/gpdb_src/src/test/regress/
+
+        make PGOPTIONS="-c optimizer=off" installcheck-resgroup || (
```
I run
```sh
IMAGE=hub.adsw.io/library/gpdb6_regress:latest bash arenadata/scripts/run_resgroup_test.bash
```
with original file `gpAux/gpdemo/demo_cluster.sh` and all tests are passed!
I think it is because of `optimizer=off`.

KnightMurloc committed:
Creating ssh keys in init_container can lead to a race condition during parallel
initialization.

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67032

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1183313

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1183314

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67335

BenderArenadata commented:
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1199337

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1199338

BenderArenadata commented:
Allure report https://allure.adsw.io/launch/67345

BenderArenadata commented:
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200113
